r2-core/src/test/java/com/linkedin/r2/util/TestServerRetryTracker.java

```diff
@@ -0,0 +1,106 @@
+/*
+   Copyright (c) 2020 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.r2.util;
+
+import com.linkedin.r2.filter.transport.ServerRetryFilter;
+import com.linkedin.util.clock.SettableClock;
+import org.testng.Assert;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+
+public class TestServerRetryTracker
+{
+  private ServerRetryTracker _serverRetryTracker;
+  private SettableClock _clock;
+
+  @BeforeMethod
+  public void setUp()
+  {
+    _clock = new SettableClock();
+    _serverRetryTracker = new ServerRetryTracker(ServerRetryFilter.DEFAULT_RETRY_LIMIT,
+        ServerRetryFilter.DEFAULT_AGGREGATED_INTERVAL_NUM, ServerRetryFilter.DEFAULT_MAX_REQUEST_RETRY_RATIO,
+        ServerRetryFilter.DEFAULT_UPDATE_INTERVAL_MS, _clock);
+  }
+
+  @Test
+  public void testEmptyServerRetryTracker()
+  {
+    for (int i = 0; i < 10; i++)
+    {
+      Assert.assertTrue(_serverRetryTracker.isBelowRetryRatio());
+      Assert.assertEquals(_serverRetryTracker.getRetryRatio(),0.0, 0.0001);
+      _clock.addDuration(ServerRetryFilter.DEFAULT_UPDATE_INTERVAL_MS);
+    }
+  }
+
+  @Test
+  public void testServerRetryTracker()
```

Comment:
It's kind of hard to separate out the multiple window scenarios because it builds upon the single window case. After separating it out in the latest commit, the test case is still pretty lengthy.